### PR TITLE
HealthBench Evaluation Task (Hard + Easy Subsets)

### PR DESCRIPTION
**Pull request by shiraeisenberg**
# Add HealthBench Evaluation Task (Hard + Easy Subsets)

## Summary

Adds **HealthBench** as a new evaluation task — the first rubric-based, LLM-as-judge eval in PostTrainBench.

Includes **two subsets**:
- **HealthBench Hard** (default): 1,000 difficult examples where base models score ~0%
- **HealthBench Easy** (new): 1,000 filtered examples targeting ~40% base model accuracy

Both subsets use physician-curated medical conversations with ~6-12 rubric criteria per example, graded by GPT-5-mini against physician-written standards.

## Why HealthBench?

Current PTB tasks have **binary pass/fail signals**:
- GSM8K: Math is right or wrong
- HumanEval: Code passes tests or doesn't
- BFCL: Function call format matches or doesn't

HealthBench introduces **multi-dimensional rubric scoring** (-10 to +10 per criterion), testing whether agents can post-train for fuzzy, open-ended objectives where success is measured by nuanced evaluation rather than exact matches.

## Baseline Results (200 samples)

| Model | Overall | Accuracy Axis | Communication Axis |
|-------|---------|---------------|-------------------|
| Qwen3-1.7B-Base | 0% | 0% | 0% |
| Qwen3-4B-Base | 0% | 1% | 3.4% |
| SmolLM3-3B-Base | 0% | 2% | **18%** |
| DeepSeek-V2-Lite | 0% | 0.1% | **17%** |
| Gemma-3-4B-pt | 0% | 2% | **14%** |
| Qwen3-1.7B (instruct) | 0% | **7.8%** | 2% |

**Post-training headroom:** ~8% accuracy improvement from base → instruct demonstrates the potential gains agents can achieve.

**Key finding:** Different base models show very different communication scores (0-18%), suggesting pre-training data/approach significantly affects this axis.

## HealthBench Easy (NEW)

For PTB 1.0, includes a filtered "Easy" subset that allows agents to show progress during post-training:

| Metric | Easy | Hard |
|--------|------|------|
| Examples | 1,000 | 1,000 |
| Avg criteria/example | 6.3 | 11.8 |
| Negative criteria/example | 1.3 | 4.1 |
| **SmolLM3-3B-Base overall** | **27.6%** | ~0% |
| **SmolLM3-3B-Base accuracy** | **40.4%** | 2% |

**Filtering criteria:**
- Non-hard examples (excluded from Hard subset)
- ≤2 negative criteria per example
- Stratified sampled to preserve theme distribution

```bash
# Easy subset (for PTB 1.0)
python evaluate.py --model-path Qwen/Qwen3-1.7B-Base --subset easy --limit 200

# Hard subset (default)
python evaluate.py --model-path Qwen/Qwen3-1.7B-Base --subset hard --limit 200
```

## Files Added

```
src/eval/tasks/healthbench/
├── benchmark.txt                    # "HealthBench Hard"
├── benchmark_easy.txt               # "HealthBench Easy"
├── evaluate.py                      # Main entry point (vLLM + LLM judge)
├── evaluation_code/
│   ├── __init__.py
│   ├── data_loader.py               # Downloads/caches HealthBench data
│   ├── grader.py                    # LLM-as-judge grading (GPT-5-mini)
│   └── scoring.py                   # Score aggregation with bootstrap stderr
└── task_context/
    └── README.md                    # Agent instructions for post-training
```

## Usage

```bash
# Quick test - Easy subset (5 examples)
python src/eval/tasks/healthbench/evaluate.py \
  --model-path Qwen/Qwen3-1.7B-Base \
  --subset easy \
  --limit 5

# Quick test - Hard subset (5 examples)
python src/eval/tasks/healthbench/evaluate.py \
  --model-path Qwen/Qwen3-1.7B-Base \
  --subset hard \
  --limit 5

# Full evaluation with output
python src/eval/tasks/healthbench/evaluate.py \
  --model-path final_model/ \
  --subset easy \
  --json-output-file results.json

# Use GPT-4.1 for paper-comparable results
python src/eval/tasks/healthbench/evaluate.py \
  --model-path final_model/ \
  --subset hard \
  --grader-model gpt-4.1 \
  --json-output-file results.json
```

## Output Format

```json
{
  "accuracy": 0.078,
  "stderr": 0.012,
  "n_examples": 200,
  "total_grader_calls": 2367,
  "by_theme": {
    "communication": 0.05,
    "emergency_referrals": 0.12,
    ...
  },
  "by_axis": {
    "accuracy": 0.078,
    "communication": 0.02,
    "context_seeking": 0.03,
    ...
  }
}
```

Primary metric is `accuracy` (overall normalized score, 0-1). Additional breakdowns by theme (7 categories) and axis (5 behavioral dimensions) provided for analysis.

## Runtime & Cost

| Configuration | Examples | Runtime | Grader Cost |
|--------------|----------|---------|-------------|
| Quick test | 5 | ~1 min | ~$0.15 |
| Dev iteration | 50 | ~3 min | ~$1.50 |
| Full eval | 200 | ~6 min | ~$6.50 |
| Complete | 1000 | ~25 min | ~$33 |

*Runtimes on H100. Grader costs using GPT-5-mini.*

## Requirements

- `OPENAI_API_KEY` environment variable (for grader)
- vLLM (for model inference)
- Python packages: `openai`, `tiktoken` (explicit in `containers/standard.def`), plus `numpy`, `requests`, `tqdm` (transitive dependencies of vLLM/pandas/transformers)

**Note:** All required packages are already in the standard PTB container. No new dependencies needed.

Data downloads automatically from OpenAI blob storage on first run and caches locally to `data/` (gitignored).

## Open Questions

1. **Templates path:** Code expects `--templates-dir` relative to repo root (default: `templates/`). Is this the correct convention?

2. **Data cache location:** Currently caches to `data/healthbench_hard.jsonl` within the task directory. Should this go elsewhere?

3. **Default limit:** Currently `None` (full 1000 examples). Should we default to a smaller number for faster iteration?

## Future Work

- **Judge comparison:** GPT-4.1 vs GPT-4o-mini vs GPT-5-mini correlation analysis (to validate cheaper graders)

## References

- [HealthBench Paper](https://arxiv.org/abs/2505.08775) — OpenAI, May 2025
- [HealthBench Data](https://openaipublic.blob.core.windows.net/simple-evals/healthbench/hard_2025-05-08-21-00-10.jsonl)

## Authors

- Shira Eisenberg
- Karina Nguyen

### CODE REVIEW

**Comment by max-andr** [comment]
Just a quick question: in the second table, the 40.4% is achieved by the base or instruct SmolLM3-3B model? I think it should be instruct, right?

**Comment by max-andr** [comment]
I guess then the key question is how quickly we can set up the infra for external contributors. @hrdkbhatnagar started working on Harbor (https://github.com/laude-institute/harbor) support, which should be much more flexible and user-friendly compared to the current infra setup. However, the last time we spoke with @hrdkbhatnagar, it wasn't clear how much more time would be needed for this. @hrdkbhatnagar - curious to hear your thoughts on this!

Alternatively, @shiraeisenberg, do you think you could help @hrdkbhatnagar with setting up Harbor (or some other user-friendly infra) for the benchmark? That would be convenient for us, since then we can allocate more time for paper writing and running various ablations.


> Question: by what date would you need the HealthBench results to include them in v1? I want to make sure we're aiming at the right target before committing to the sprint.

I would say the latest is 1 week before the paper deadline which is on Jan 28. So then it would be by Jan 21.

**Comment by rank-and-file** [comment]
Hi, thanks for adding this, I think having HealthBench in PostTrainBench would be great!

Some comments:
- did you compute the scores for the templates provided in src/eval/templates (and for qwen the standard template)? -> these are the ones we use for evaluating. You can supply them by using the `--templates-dir` option in the `evaluate.py` script.
- we can remove the "hard" subset support entirely. This would simplify the integration. So no `--subset` option and having only one `benchmark.txt` with "HealthBench (Easy)" or similar.
- The goal for HealthBench Easy would be to have a subset of questions such that the instruction tuned versions of the models perform significantly better than the base models. It is not necessary that base-models themselves perform well. Because this will show that the benchmark is feasible even for smaller (instruction tuned) models.
- If you can provide a list of huggingface datasets which the models would typically use when training for HealthBench, this would be great. Then we can add them to our huggingface cache downloading in `containers/download_hf_cache`. This can be obtained by asking the 3 agents for very inclusive lists of benchmarks which they would consider using when training for HealthBench. Of course very large datasets are not needed (say above 15G is too large). You can add them to `containers/download_hf_cache/resources.json`

Maybe if we get this done well, it is possible to do experiments from our end in time :)
I think proving Harbor support will be difficult to do in time.

**Comment by hrdkbhatnagar** [comment]
Hey, thanks for the updates!

**Re: running PTB on your end (@shiraeisenberg):** I'm about 90% done building a [Harbor](https://harborframework.com/) adapter for PTB. Currently testing for feature parity with the existing codebase and ironing out some bugs with Modal (for cloud compute). Once the adapter is complete, I can onboard you so you can start running PTB independently. Or I could already show you the current codebase and you could start building on top of it. 

**Re: timeline:** Given the Jan 21 deadline for results, I'll prioritize finishing the adapter by today/tomorrow.

**Comment by shiraeisenberg** [comment]
Thanks for the detailed feedback! To address each point:
	1.	Templates: Will rerun with --templates-dir pointing to src/eval/templates. Will update results.
	2.	Remove Hard subset: Makes sense. I’ll simplify to just HealthBench Easy and update benchmark.txt accordingly.
	3.	HuggingFace datasets: I’ll compile a list of datasets agents would likely use for medical/health training and add to resources.json. Will have this ready by our meeting tomorrow.
	4.	Harbor onboarding: Happy to connect with @hrdkbhatnagar once the adapter is ready.
What’s the best way to coordinate on Slack? I'm still traveling but get home ~10PM EST tonight, so I will run stuff then or in the morning.

**Comment by shiraeisenberg** [comment]
## Update: Templates + Simplified to Easy Only + Base vs Instruct Comparison

### 1. Templates Support
Reran validation using `--templates-dir src/eval/templates/` with the official PTB SmolLM template (`smollm.jinja`).

### 2. Simplified to Easy Only
Removed the `--subset` option entirely:
- Single `benchmark.txt` containing "HealthBench"
- Data loader only loads Easy subset by default
- Cleaned up all Hard-related code paths

### 3. Base vs Instruct Comparison (50 samples, PTB templates, gpt-5-mini)

| Metric | SmolLM3-3B-Base | SmolLM3-3B (Instruct) | Gap |
|--------|-----------------|----------------------|-----|
| **Overall** | 27.2% | **50.3%** | +23.1% ✅ |
| Accuracy | 63.9% | 50.8% | -13.1% |
| Context Awareness | 25.9% | 50.3% | +24.4% ✅ |
| Completeness | 5.1% | 37.5% | +32.4% ✅ |
| Communication | 0% | **100%** | +100% ✅ |

**Observations:**
- **Overall gap exists**: 27% → 50% shows meaningful improvement from post-training
- **Communication/Completeness**: Instruct model is dramatically better
- **Accuracy paradox**: Base model scores *higher* on accuracy axis (64% vs 51%)

The base model's high accuracy but poor communication/completeness suggests it gives factually correct but incomplete/poorly formatted responses. The instruct model provides more complete, well-communicated answers.

### Discussion/Question: Is the ~23% overall gap sufficient?

The current Easy subset shows:
- ✅ Clear overall improvement from base → instruct
- ✅ Dramatic gains on communication/completeness axes
- ⚠️ Base model accuracy is higher than expected (64%)

**Options to consider:**

1. **Accept current results** — The 23% overall gap is meaningful, and the per-axis breakdown shows post-training clearly helps on communication/completeness. The fact that base models get some facts right on "easy" questions isn't necessarily problematic.

2. **Tighten Easy subset filtering** — Could add additional filters (e.g., require multi-turn conversations, minimum criteria count) to increase difficulty while staying in the "easy" range.

3. **Test with other model pairs** — Run the same comparison on Qwen or Gemma to see if the pattern holds.

What's your preference? Happy to iterate on the subset selection if needed before the Jan 21 deadline.

### 4. HuggingFace Datasets
Still compiling the list — will have this ready shortly.

### Results Files
- [healthbench_easy_smollm3_ptb_templates.json](https://github.com/user-attachments/files/24575153/healthbench_easy_smollm3_ptb_templates.json)
- [healthbench_easy_smollm3_instruct_ptb_templates.json](https://github.com/user-attachments/files/24575152/healthbench_easy_smollm3_instruct_ptb_templates.json)
